### PR TITLE
Feature/detalizing logs

### PR DESCRIPTION
**Pull request by androndo**
# PR Summary

Detalizing logs:
 * changed old fields from message text (like [trigger_id], etc) to structured fileds
 * api: added http-fields
 * notifier: added configuring log level by subscription and contact
 * checker: added configuring log level by trigger
 * docs changes [here](https://github.com/moira-alert/doc/pull/74)
Relates #441

### CODE REVIEW

File: logging/zerolog_adapter/logger.go [hunk 1]
```diff
@@ -72,43 +72,83 @@ func getLogWriter(logFileName string) (io.Writer, error) {
 }
 
 func (l Logger) Debug(args ...interface{}) {
-	l.Logger.Debug().Timestamp().Msg(fmt.Sprint(args...))
+	event := l.Logger.Debug()
+	if event == nil {
+		return
+	}
```
Comment (litleleprikon) [inline logging/zerolog_adapter/logger.go:78]: Which circumstance can lead to event to be nil?

File: notifier/events/event.go [hunk 2]
```diff
@@ -140,15 +148,17 @@ func (worker *FetchEventsWorker) processEvent(event moira.NotificationEvent) err
...
 		sub, err := worker.Database.GetSubscription(*event.SubscriptionID)
 		if err != nil {
 			worker.Metrics.SubsMalformed.Mark(1)
 			return nil, fmt.Errorf("error while read subscription %s: %s", *event.SubscriptionID, err.Error())
 		}
 		return &sub, nil
 	} else if event.ContactID != "" {
-		worker.Logger.Debugf("Getting contactID %s for test message", event.ContactID)
+		log := worker.Logger.Clone().String(moira.LogFieldNameContactID, event.ContactID)
+		log.Debugf("Getting contactID %s for test message", event.ContactID)
```
Comment (litleleprikon) [inline notifier/events/event.go:161]: Again, there is no need to format contact id to message

File: notifier/events/event.go [hunk 3]
```diff
@@ -175,11 +185,15 @@ func (worker *FetchEventsWorker) isNotificationRequired(subscription *moira.Subs
 	}
 	if event.State != moira.StateTEST {
 		if !subscription.Enabled {
-			worker.Logger.Debugf("Subscription %s is disabled", subscription.ID)
+			worker.Logger.Clone().String(moira.LogFieldNameSubscriptionID, subscription.ID).
+				String(moira.LogFieldNameTriggerID, trigger.ID).
+				Debugf("Subscription %s is disabled", subscription.ID)
```
Comment (litleleprikon) [inline notifier/events/event.go:190]: No need to format subscription id to message

File: notifier/events/event.go [hunk 4]
```diff
@@ -175,11 +185,15 @@ func (worker *FetchEventsWorker) isNotificationRequired(subscription *moira.Subs
...
 			return false
 		}
 		if subscription.MustIgnore(&event) {
-			worker.Logger.Debugf("Subscription %s is managed to ignore %s -> %s transitions", subscription.ID, event.OldState, event.State)
+			worker.Logger.Clone().String(moira.LogFieldNameSubscriptionID, subscription.ID).
+				String(moira.LogFieldNameTriggerID, trigger.ID).
+				Debugf("Subscription %s is managed to ignore %s -> %s transitions", subscription.ID, event.OldState, event.State)
```
Comment (litleleprikon) [inline notifier/events/event.go:196]: Let's move subscription id and states to structured fields

File: notifier/plotting.go [hunk 5]
```diff
@@ -85,7 +86,9 @@ func resolveMetricsWindow(logger moira.Logger, trigger moira.TriggerData, pkg No
 	// try to resolve package window, force default realtime window on fail for both local and remote triggers
 	from, to, err := pkg.GetWindow()
 	if err != nil {
-		logger.Warningf("Failed to get trigger %s package window: %s, using default %s window", trigger.ID, err.Error(), defaultTimeRange.String())
+		getLogWithPackageContext(&logger, &pkg).
+			Warningf("Failed to get trigger %s package window: %s, using default %s window",
```
Comment (litleleprikon) [inline notifier/plotting.go:90]: I think that trigger id can be moved to structured fields

File: notifier/scheduler.go [hunk 6]
```diff
@@ -63,10 +63,12 @@ func (scheduler *StandardScheduler) ScheduleNotification(now time.Time, event mo
...
 	}
-	scheduler.logger.Debugf(
-		"Scheduled notification for contact %s:%s trigger %s at %s (%d)",
-		contact.Type, contact.Value, trigger.Name,
-		next.Format("2006/01/02 15:04:05"), next.Unix())
+	scheduler.logger.Clone().
+		String(moira.LogFieldNameContactID, contact.ID).
+		String(moira.LogFieldNameTriggerID, trigger.ID).
+		Debugf("Scheduled notification for contact %s:%s trigger %s at %s (%d)",
+			contact.Type, contact.Value, trigger.Name,
```
Comment (litleleprikon) [inline notifier/scheduler.go:70]: Let's move value, type and trigger name to structured fields

File: notifier/scheduler.go [hunk 7]
```diff
@@ -92,13 +94,15 @@ func (scheduler *StandardScheduler) calculateNextDelivery(now time.Time, event *
 	subscription, err := scheduler.database.GetSubscription(moira.UseString(event.SubscriptionID))
 	if err != nil {
 		scheduler.metrics.SubsMalformed.Mark(1)
-		scheduler.logger.Debugf("Failed get subscription by id: %s. %s", moira.UseString(event.SubscriptionID), err.Error())
+		getLogWithEventContext(&scheduler.logger, event).
+			Debugf("Failed get subscription by id: %s. %s", moira.UseString(event.SubscriptionID), err.Error())
```
Comment (litleleprikon) [inline notifier/scheduler.go:98]: Subscription id here can be moved to structured fields

File: notifier/scheduler.go [hunk 8]
```diff
@@ -92,13 +94,15 @@ func (scheduler *StandardScheduler) calculateNextDelivery(now time.Time, event *
...
 		return next, alarmFatigue
 	}
 
 	if subscription.ThrottlingEnabled {
 		if next.After(now) {
-			scheduler.logger.Debugf("Using existing throttling for trigger %s: %s", event.TriggerID, next)
+			getLogWithEventContext(&scheduler.logger, event).
+				Debugf("Using existing throttling for trigger %s: %s", event.TriggerID, next)
```
Comment (litleleprikon) [inline notifier/scheduler.go:105]: Trigger id here can be moved to structured fields

File: notifier/scheduler.go [hunk 9]
```diff
@@ -108,10 +112,12 @@ func (scheduler *StandardScheduler) calculateNextDelivery(now time.Time, event *
 				count := scheduler.database.GetNotificationEventCount(event.TriggerID, from.Unix())
 				if count >= level.count {
 					next = now.Add(level.delay)
-					scheduler.logger.Debugf("Trigger %s switched %d times in last %s, delaying next notification for %s",
-						event.TriggerID, count, level.duration, level.delay)
+					getLogWithEventContext(&scheduler.logger, event).
+						Debugf("Trigger %s switched %d times in last %s, delaying next notification for %s",
```
Comment (litleleprikon) [inline notifier/scheduler.go:116]: Let's move trigger id to structured fields

File: notifier/scheduler.go [hunk 10]
```diff
@@ -125,7 +131,8 @@ func (scheduler *StandardScheduler) calculateNextDelivery(now time.Time, event *
 	}
 	next, err = calculateNextDelivery(&subscription.Schedule, next)
 	if err != nil {
-		scheduler.logger.Errorf("Failed to apply schedule for subscriptionID: %s. %s.", moira.UseString(event.SubscriptionID), err)
+		getLogWithEventContext(&scheduler.logger, event).
+			Errorf("Failed to apply schedule for subscriptionID: %s. %s.", moira.UseString(event.SubscriptionID), err)
```
Comment (litleleprikon) [inline notifier/scheduler.go:135]: Let's move subscription id to structured fields

**Review by litleleprikon — CHANGES_REQUESTED** [review]

File: api/middleware/logger.go [hunk 11]
```diff
@@ -79,19 +77,37 @@ func newLogEntry(logger moira.Logger, request *http.Request) *apiLoggerEntry {
...
+	log.String("username", userName)
+
+	entry.buf.WriteString("\"")
+	fmt.Fprintf(entry.buf, "%s ", request.Method)
+	fmt.Fprintf(entry.buf, "%s %s\"", uri, request.Proto)
 	entry.buf.WriteString(" from ")
 	entry.buf.WriteString(request.RemoteAddr)
 	entry.buf.WriteString(" by ")
 	entry.buf.WriteString(userName)
 	entry.buf.WriteString(" - ")
```
[the remark this answers is not in the document]
Comment (androndo) [reply api/middleware/logger.go:97]: Already done, some lines above used log.String to fill its.

File: logging/zerolog_adapter/logger.go (hunk 1 above)
Comment (androndo) [reply logging/zerolog_adapter/logger.go:78]: zerolog returns `event = nil` when log level is disabled, and do nothing on all next calls `event.Fileds(...).Timestamp().Msg(...)`

File: checker/check.go [hunk 12]
```diff
@@ -315,7 +317,8 @@ func (triggerChecker *TriggerChecker) getMetricStepsStates(metricName string, me
 	}
 
 	checkPoint := last.GetCheckPoint(checkPointGap)
-	triggerChecker.logger.Debugf("[TriggerID:%s][MetricName:%s] Checkpoint: %v", triggerChecker.triggerID, metricName, checkPoint)
+	triggerChecker.logger.Clone().String(moira.LogFieldNameMetricName, metricName).
+		Debugf("Checkpoint: %v", checkPoint)
```
[the remark this answers is not in the document]
Comment (androndo) [reply checker/check.go]: Added Int64 method to Logger interface

[the remark this answers is not in the document]
Comment (androndo) [reply checker/check.go]: Added `Int64` method to Logger interface

**Review by litleleprikon — DISMISSED** [review]

**Review by litleleprikon — APPROVED** [review]

**Pull request merged**